### PR TITLE
Manually approve new registration

### PR DESCRIPTION
**Pull request by ghost**
Add manual confirmation of the newly registered users.

This change speeds up the process of adding new confirmed users for Gitea instances without an external auth mechanism. (Currently, the option is to manually create a new user by admin.)

As this is a small change, I haven't created an issue. However, I am ready to do that if it is desired.

I am sure that the PR is not perfect and looking forward to the comments. Thanks!

### CODE REVIEW

File: routers/user/auth.go [hunk 1]
```diff
@@ -949,7 +949,7 @@ func LinkAccountPostRegister(ctx *context.Context, cpt *captcha.Captcha, form au
 		Name:        form.UserName,
 		Email:       form.Email,
 		Passwd:      form.Password,
-		IsActive:    !setting.Service.RegisterEmailConfirm,
+		IsActive:    !setting.Service.RegisterEmailConfirm || !setting.Service.RegisterManualConfirm,
```
Comment (lafriks) [inline routers/user/auth.go]: This will break `RegisterEmailConfirm` feature with `RegisterManualConfirm` disabled

File: routers/user/auth.go [hunk 2]
```diff
@@ -1141,7 +1141,7 @@ func SignUpPost(ctx *context.Context, cpt *captcha.Captcha, form auth.RegisterFo
 		Name:     form.UserName,
 		Email:    form.Email,
 		Passwd:   form.Password,
-		IsActive: !setting.Service.RegisterEmailConfirm,
+		IsActive: !setting.Service.RegisterEmailConfirm || !setting.Service.RegisterManualConfirm,
```
Comment (lafriks) [inline routers/user/auth.go]: same here

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/user/auth.go (hunk 1 above)
Comment (ghost) [reply routers/user/auth.go]: My bad. Is
```
IsActive:    !(setting.Service.RegisterEmailConfirm || setting.Service.RegisterManualConfirm),
```
better? I think it should work already.

**Comment by lafriks** [comment]
But if you confirm by email user will still be activated so it will not be functional with both enabled

**Comment by qeef** [comment]
That's true. It would be possible to choose (exclusively) between confirming automatically, by email, or manually. (Is that problem?)

I think that enabling more than one way of confirmation would need database change but maybe I am wrong?

**Comment by lafriks** [comment]
Yeah it would but than you should add a note for this in docs and example ini that this feature is supported only when `REGISTER_EMAIL_CONFIRM` is disabled.

File: docs/content/doc/advanced/config-cheat-sheet.zh-cn.md [hunk 3]
```diff
@@ -122,6 +122,7 @@ menu:
 - `ACTIVE_CODE_LIVE_MINUTES`: 登录验证码失效时间，单位分钟。
 - `RESET_PASSWD_CODE_LIVE_MINUTES`: 重置密码失效时间，单位分钟。
 - `REGISTER_EMAIL_CONFIRM`: 启用注册邮件激活，前提是 `Mailer` 已经启用。
+- `REGISTER_MANUAL_CONFIRM`: **false**: Enable this to manually confirm new registrations. Requires `REGISTER_EMAIL_CONFIRM` to be disabled.
```
Comment (lafriks) [inline docs/content/doc/advanced/config-cheat-sheet.zh-cn.md]: @lunny can you translate this?

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
I think we have to rethink this more. Email confirmation will let you know the emails are addressable but manually confimation will not. I think we have more codes need to be changed on those aspect.

**Comment by qeef** [comment]
> I think we have to rethink this more. Email confirmation will let you know the emails are addressable but manually confimation will not. I think we have more codes need to be changed on those aspect.

That's true, @lunny. However, this PR is for a particular use-case: We are a small team at university and we need to let students and cooperating partners join our development. We have some open source project, some of them are closed. We don't even need to track email addresses.

The point is that if we want to control access to the gitea instance, we can currently use external authentication (we don't have,) or manually create users. This simple PR will improve our workflow by just approving new users. Backward compatibility is preserved.

**Comment by lunny** [comment]
@qeef I know. I don't against the PR. I mean, if we can mark these email addresses as non-confirmed so that the future PRs could send a confirmation email to know if the notification emails could be received.

**Comment by qeef** [comment]
@lunny Marking emails as non-confirmed is one option, IMO. It would require a new column in the database.

The second option that comes into my mind is adding two new columns `needs_confirm` and `is_confirm` which would work as flags. In gitea configuration, you set `needs_confirm` flags, like email approved, manual approved, or anything else. When a particular approvement is done, you update `is_confirm` flags. And finally, when `needs_confirm == is_confirm`, you set `is_active` to `true`. However, changes in database and gui are inevitable and I am not sure about profitability.

Did you have a particular idea?

**Comment by qeef** [comment]
I was yet thinking about *future PRs* as mentioned above. I tend to say that email confirmation is a little bit of different functionality.

Currently, `is_active` does not mean the notification emails could be received anyway. Moreover, this PR does not limit future PRs, in my opinion.

Therefore, I propose to discuss emal confirmation in different issue/PR.

**Comment by zeripath** [comment]
Hi @qeef,

is_active and is_confirmed would be essentially the same.

is_active exists to prevent users from pretending to be someone else and prevent that person from logging in.

I'm don't quite understand how this is supposed to work.

How are the users registering? what are they doing?

**Comment by lafriks** [comment]
@zeripath for this PR use case is that if server does not have mail sending set up and user registers it is automatically activated but @qeef wants that users registers but admin than activates it manually

**Comment by zeripath** [comment]
can't they just use the dummy mailer?

**Comment by qeef** [comment]
Hi @zeripath.

> is_active exists to prevent users from pretending to be someone else and prevent that person from logging in.

it means that `is_active` is used for authentication.

> I'm don't quite understand how this is supposed to work.

This PR use `is_active` column for authorization when no authentication over email is required. The backward compatibility is ensured.

The use-case is the manual control of the access to the Gitea instance (without necessity of external authentication configuration.) Currently, the manual access control can be done by creating users by admins. This PR let users create their accounts themselves, and let administrators just approve it (therefore speeds up the process.)

**Comment by qeef** [comment]
Guys, is there anything more I can do?

File: modules/setting/service.go [hunk 4]
```diff
@@ -17,6 +17,7 @@ var Service struct {
 	ActiveCodeLives                         int
 	ResetPwdCodeLives                       int
 	RegisterEmailConfirm                    bool
+	RegisterManualConfirm                   bool
```
Comment (a1012112796) [inline modules/setting/service.go:20]: need set the value in ``newService()`` (see https://github.com/go-gitea/gitea/pull/13083/files#diff-0ab2a298fe8b3109330b3679b3ae75683af157d2e1123251f21e4bf6ed0c08efR61 )

File: docs/content/doc/advanced/config-cheat-sheet.zh-cn.md (hunk 3 above)
Comment (a1012112796) [reply docs/content/doc/advanced/config-cheat-sheet.zh-cn.md]:
```suggestion
- `REGISTER_MANUAL_CONFIRM`: **false**: 新注册用户必须由管理员手动激活,启用此选项需取消`REGISTER_EMAIL_CONFIRM`.
```

File: modules/setting/service.go (hunk 4 above)
Comment (ghost) [reply modules/setting/service.go:20]: Sorry, I am a bit lost. The `RegisterManualConfirm` follows the implementation of `RegisterEmailConfirm` and `RegisterEmailConfirm` is not set in `newService()` neither.

Comment (a1012112796) [reply modules/setting/service.go:20]: it's setting in here
https://github.com/go-gitea/gitea/blob/63aeb0a859499623becc1d1e7c8a2ad57439e139/modules/setting/mailer.go#L112-L121

Comment (ghost) [reply modules/setting/service.go:20]: Oh, I got it. Should be solved now (https://github.com/go-gitea/gitea/pull/13083/commits/95a3b174d0ae85559f97192ebe7e02c1ee43d9c4#diff-0ab2a298fe8b3109330b3679b3ae75683af157d2e1123251f21e4bf6ed0c08efR67)

**Comment by qeef** [comment]
I fixed the last review and rebased to the current `master`.

**Review by lunny — APPROVED** [review]

**Pull request merged**